arch/arc/core/mpu/arc_mpu_v6_internal.h

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+#define ZEPHYR_ARCH_ARC_CORE_MPU_ARC_MPU_V6_INTERNAL_H_
+
+#define AUX_MPU_RDB_VALID_MASK (0x1)
+#define AUX_MPU_EN_ENABLE   (0x40000000)
+#define AUX_MPU_EN_DISABLE  (0xBFFFFFFF)
+#define AUX_MPU_EN_BANK_MASK (0x1)
+
+#define AUX_MPU_RDP_REGION_SIZE(bits)   (((bits - 1) & 0x3) | (((bits - 1) & 0x1C) << 7))
+
+#define AUX_MPU_RDP_ATTR_MASK (0x1F8)
+#define AUX_MPU_RDP_SIZE_MASK (0xE03)
+
+/* For MPU version 6, the minimum protection region size is 32 bytes */
+#define ARC_FEATURE_MPU_ALIGNMENT_BITS 5
+#define ARC_FEATURE_MPU_BANK_SIZE      16
+
+/**
+ * This internal function select a MPU bank
+ */
+static inline void _bank_select(uint32_t bank)
+{
+	uint32_t val;
+
+	val = z_arc_v2_aux_reg_read(_ARC_V2_MPU_EN) & (~AUX_MPU_EN_BANK_MASK);
+	z_arc_v2_aux_reg_write(_ARC_V2_MPU_EN, val | bank);
+}
+/**
+ * This internal function initializes a MPU region
+ */
+static inline void _region_init(uint32_t index, uint32_t region_addr,
+				uint32_t size, uint32_t region_attr)
+{
+	uint32_t bank = index / ARC_FEATURE_MPU_BANK_SIZE;
+
+	index = (index % ARC_FEATURE_MPU_BANK_SIZE) * 2U;
+
+	if (size > 0) {
+		uint8_t bits = find_msb_set(size) - 1;
+
+		if (bits < ARC_FEATURE_MPU_ALIGNMENT_BITS) {
+			bits = ARC_FEATURE_MPU_ALIGNMENT_BITS;
+		}
+
+		if ((1 << bits) < size) {
+			bits++;
+		}
+
+		region_attr &= ~(AUX_MPU_RDP_SIZE_MASK);
+		region_attr |= AUX_MPU_RDP_REGION_SIZE(bits);
+		region_addr |= AUX_MPU_RDB_VALID_MASK;
+	} else {
+		region_addr = 0U;
+	}
+
+	_bank_select(bank);
+	z_arc_v2_aux_reg_write(_ARC_V2_MPU_RDP0 + index, region_attr);
+	z_arc_v2_aux_reg_write(_ARC_V2_MPU_RDB0 + index, region_addr);
+}
+
+/**
+ * This internal function is utilized by the MPU driver to parse the intent
+ * type (i.e. THREAD_STACK_REGION) and return the correct region index.
+ */
+static inline int get_region_index_by_type(uint32_t type)
+{
+	/*
+	 * The new MPU regions are allocated per type after the statically
+	 * configured regions. The type is one-indexed rather than
+	 * zero-indexed.
+	 *
+	 * For ARC MPU v6, the smaller index has higher priority, so the
+	 * index is allocated in reverse order. Static regions start from
+	 * the biggest index, then thread related regions.
+	 *
+	 */
+	switch (type) {
+	case THREAD_STACK_USER_REGION:
+		return get_num_regions() - mpu_config.num_regions
+		       - THREAD_STACK_REGION;
```

Comment:
We don't exceed the Zephyr line limit here (100 symbols per line), please keep it in one line.

Please check rest of the code - we have a lot of places where we can put the whole expression in one line. It improves code readability a lot.